src/GafferScene/Instancer.cpp

```diff
@@ -2727,29 +2749,57 @@ void Instancer::InstancerCapsule::render( IECoreScenePreview::Renderer *renderer
 			std::string name;
 			IECoreScenePreview::Renderer::AttributesInterfacePtr attribsStorage;
 
+			Context::EditableScope prototypeScope( scope.context() );
+
 			for( size_t idx = r.begin(); idx != r.end(); ++idx )
 			{
 				int pointIndex = pointIndicesForPrototype[idx];
 
-				int prototypeIndex = 0;
-				if( prototypeIndices.size() )
+				const Prototype *proto;
+				if( constantPrototype )
 				{
-					prototypeIndex = prototypeIndices[ idx ];
+					proto = constantPrototype.get();
+				}
+				else
+				{
+					// The prototype depends on the context, so we need to find the prototype context for
+					// this instance.
+
+
+					// We find the capsules using the engine at shutter open, but the time used to construct the capsules
+					// must be the on-frame time, since the capsules will add their own shutter ( and we also handle
+					// the shutter ourselves for transform matrices )
+					//
+					// For most context variables, we are overwriting them for each prototype anyway, so
+					// we can reuse the context. But timeOffset is relative, so it's important that we reset the
+					// time before we do setPrototypeContextVariables for the next element. ( Should this be more
+					// general instead of assuming that frame is the only variable for which offsetMode may be set? )
+					prototypeScope.setFrame( onFrameTime );
+
+					engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
+
+					proto = prototypeCache.get( PrototypeCacheGetterKey( prototypeScope.context() ) ).get();
+				}
+
+				if( !proto->m_object.size() )
+				{
+					// No object to render. This could happen if the protype didn't meet the
+					// RenderOptions::purposeIncluded test.
+					continue;
 				}
-				Prototype &proto = prototypes[prototypeIndex];
 
 				IECoreScenePreview::Renderer::AttributesInterface *attribs;
 				int instanceId = engines[0]->instanceId( pointIndex );
 				if( hasAttributes )
 				{
-					CompoundObjectPtr currentAttributes = proto.attributes->copy();
+					CompoundObjectPtr currentAttributes = proto->m_attributes->copy();
```

Comment:
This call to `copy()` actually showed up as relatively costly in some profiling I did the other day on a production instancing setup. It's doing a deep copy, and although each of the individual data items will do lazy-copy-on-write, the tracking machinery in CopyContext actually still hurts a fair bit. In this case, a simple shallow copy is sufficient. You can see examples of that in most of our attribute-processing nodes - here's [one](https://github.com/GafferHQ/gaffer/blob/main/src/GafferScene/Attributes.cpp#L217-L222).

Reply:
Good point.